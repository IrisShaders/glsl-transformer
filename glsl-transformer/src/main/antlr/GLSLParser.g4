/* GLSL grammar that extends the graphicsfuzz grammar
but is specifically built for this project*/

/*
 * Copyright 2018 The GraphicsFuzz Project Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

parser grammar GLSLParser;

@header {
import io.github.douira.glsl_transformer.parser.ExtendedParser;
}

options {
	tokenVocab = GLSLLexer;
	superClass = ExtendedParser;
}

//the root rule
translationUnit: versionStatement? externalDeclaration* EOF;

//allows for EXT_null_initializer
versionStatement:
	NR NR_VERSION version = (
		NR_GLSL_110
		| NR_GLSL_120
		| NR_GLSLES_100
		| NR_GLSL_130
		| NR_GLSL_140
		| NR_GLSL_150
		| NR_GLSL_330
		| NR_GLSLES_300
		| NR_GLSLES_310
		| NR_GLSLES_320
		| NR_GLSL_400
		| NR_GLSL_410
		| NR_GLSL_420
		| NR_GLSL_430
		| NR_GLSL_440
		| NR_GLSL_450
		| NR_GLSL_460
	) profile = (NR_CORE | NR_COMPATIBILITY | NR_ES)? NR_EOL;

externalDeclaration:
	functionDefinition
	| declaration
	| pragmaStatement
	| extensionStatement
	| customDirectiveStatement
	| includeStatement
	| layoutDefaults
	| emptyDeclaration;

emptyDeclaration: SEMICOLON;

pragmaStatement:
	NR NR_PRAGMA stdGL = NR_STDGL? (
		type = NR_IDENTIFIER
		| type = (NR_PRAGMA_DEBUG | NR_PRAGMA_OPTIMIZE) NR_LPAREN state = (
			NR_ON
			| NR_OFF
		) NR_RPAREN
		| type = NR_PRAGMA_INVARIANT NR_LPAREN state = NR_ALL NR_RPAREN
	) NR_EOL;

extensionStatement:
	NR NR_EXTENSION extensionName = NR_IDENTIFIER (
		NR_COLON extensionBehavior = (
			NR_REQUIRE
			| NR_ENABLE
			| NR_WARN
			| NR_DISABLE
		)
	)? NR_EOL;

customDirectiveStatement: NR NR_CUSTOM content = C_CONTENT? C_EOL;

includeStatement:
	NR NR_INCLUDE (
		NR_STRING_START content = S_CONTENT? S_STRING_END
		| angleStart = NR_STRING_START_ANGLE content = S_CONTENT_ANGLE? S_STRING_END_ANGLE
	) NR_EOL;

layoutDefaults:
	layoutQualifier layoutMode = (UNIFORM | IN | OUT | BUFFER) SEMICOLON;

functionDefinition: functionPrototype compoundStatement;

//Note: diverges from the spec by explicity adding a method call instead of handling it through postfixExpression in functionIdentifier
expression:
	IDENTIFIER # referenceExpression
	| (
		INT16CONSTANT
		| UINT16CONSTANT
		| INT32CONSTANT
		| UINT32CONSTANT
		| INT64CONSTANT
		| UINT64CONSTANT
		| FLOAT16CONSTANT
		| FLOAT32CONSTANT
		| FLOAT64CONSTANT
		| BOOLCONSTANT
	)																													# literalExpression
	| LPAREN value = expression RPAREN												# groupingExpression
	| left = expression LBRACKET right = expression RBRACKET	# arrayAccessExpression
	| operand = expression DOT_LENGTH_METHOD_CALL							# lengthAccessExpression
	//Note: diverges from the spec by not allowing a prefixExpression as an identifier
	//array-type function identfiers are handled by typeSpecifier
	| (IDENTIFIER | typeSpecifier) LPAREN (
		| VOID
		| parameters += expression (COMMA parameters += expression)*
	) RPAREN																				# functionCallExpression
	| operand = expression DOT member = IDENTIFIER	# memberAccessExpression
	| operand = expression op = (INC_OP | DEC_OP)		# postfixExpression
	| <assoc = right> op = (
		INC_OP
		| DEC_OP
		| PLUS_OP
		| MINUS_OP
		| LOGICAL_NOT_OP
		| BITWISE_NEG_OP
	) operand = expression																											# prefixExpression
	| left = expression op = (TIMES_OP | DIV_OP | MOD_OP) right = expression		# multiplicativeExpression
	| left = expression op = (PLUS_OP | MINUS_OP) right = expression						# additiveExpression
	| left = expression op = (LEFT_OP | RIGHT_OP) right = expression						# shiftExpression
	| left = expression op = (LT_OP | GT_OP | LE_OP | GE_OP) right = expression	# relationalExpression
	| left = expression op = (EQ_OP | NE_OP) right = expression									# equalityExpression
	| left = expression op = BITWISE_AND_OP right = expression									# bitwiseAndExpression
	| left = expression op = BITWISE_XOR_OP right = expression									# bitwiseExclusiveOrExpression
	| left = expression op = BITWISE_OR_OP right = expression										# bitwiseInclusiveOrExpression
	| left = expression op = LOGICAL_AND_OP right = expression									# logicalAndExpression
	| left = expression op = LOGICAL_XOR_OP right = expression									# logicalExclusiveOrExpression
	| left = expression op = LOGICAL_OR_OP right = expression										# logicalInclusiveOrExpression
	| <assoc = right> condition = expression QUERY_OP trueAlternative = expression COLON falseAlternative =
		expression # conditionalExpression
	| <assoc = right> left = expression op = (
		ASSIGN_OP
		| MUL_ASSIGN
		| DIV_ASSIGN
		| MOD_ASSIGN
		| ADD_ASSIGN
		| SUB_ASSIGN
		| LEFT_ASSIGN
		| RIGHT_ASSIGN
		| AND_ASSIGN
		| XOR_ASSIGN
		| OR_ASSIGN
	) right = expression													# assignmentExpression
	| left = expression COMMA right = expression	# sequenceExpression;

declaration:
	functionPrototype SEMICOLON # functionDeclaration
	| typeQualifier blockName = IDENTIFIER structBody (
		variableName = IDENTIFIER arraySpecifier?
	)? SEMICOLON # interfaceBlockDeclaration
	| typeQualifier (
		variableNames += IDENTIFIER (COMMA variableNames += IDENTIFIER)*
	)? SEMICOLON																						# variableDeclaration
	| PRECISION precisionQualifier typeSpecifier SEMICOLON	# precisionDeclaration
	| fullySpecifiedType (
		declarationMembers += declarationMember (
			COMMA declarationMembers += declarationMember
		)*
	)? SEMICOLON # typeAndInitDeclaration;

//allows for EXT_subgroup_uniform_control_flow
functionPrototype:
	attribute? fullySpecifiedType IDENTIFIER LPAREN functionParameterList RPAREN attribute?;

functionParameterList: (
		parameters += parameterDeclaration (
			COMMA parameters += parameterDeclaration
		)*
	)?;

parameterDeclaration:
	fullySpecifiedType (parameterName = IDENTIFIER arraySpecifier?)?;

//part of GL_EXT_control_flow_attributes
attribute:
	LBRACKET LBRACKET attributes += singleAttribute (
		COMMA attributes += singleAttribute
	)* RBRACKET RBRACKET;

singleAttribute:
	(prefix = IDENTIFIER COLON COLON)? name = IDENTIFIER (
		LPAREN content = expression RPAREN
	)?;

declarationMember: IDENTIFIER arraySpecifier? (ASSIGN_OP initializer)?;

fullySpecifiedType: typeQualifier? typeSpecifier;

//allows for KHR_memory_scope_semantics
storageQualifier:
	CONST
	| IN
	| OUT
	| INOUT
	| CENTROID
	| PATCH
	| SAMPLE
	| UNIFORM
	| VARYING
	| ATTRIBUTE
	| BUFFER
	| SHARED
	| RESTRICT
	| VOLATILE
	| COHERENT
	| READONLY
	| WRITEONLY
	| SUBROUTINE (
		//TYPE_NAME instead of IDENTIFIER in the spec
		LPAREN typeNames += IDENTIFIER (COMMA typeNames += IDENTIFIER)* RPAREN
	)?
	| DEVICECOHERENT
	| QUEUEFAMILYCOHERENT
	| WORKGROUPCOHERENT
	| SUBGROUPCOHERENT
	| NONPRIVATE;

layoutQualifier:
	LAYOUT LPAREN layoutQualifiers += layoutQualifierId (
		COMMA layoutQualifiers += layoutQualifierId
	)* RPAREN;

//constant expression
layoutQualifierId:
	IDENTIFIER (ASSIGN_OP expression)?	# namedLayoutQualifier
	| SHARED														# sharedLayoutQualifier;

precisionQualifier: HIGHP | MEDIUMP | LOWP;

interpolationQualifier: SMOOTH | FLAT | NOPERSPECTIVE;

invariantQualifier: INVARIANT;

preciseQualifier: PRECISE;

typeQualifier: (
		storageQualifier
		| layoutQualifier
		| precisionQualifier
		| interpolationQualifier
		| invariantQualifier
		| preciseQualifier
	)+;

//TYPE_NAME instead of IDENTIFIER in the spec
typeSpecifier: (
		IDENTIFIER
		| builtinTypeSpecifierFixed
		| builtinTypeSpecifierParseable
		| structSpecifier
	) arraySpecifier?;

//needs duplicated rule parts like this or it becomes mutually left-recursive
//constant expressions
arraySpecifier: arraySpecifierSegment+;
arraySpecifierSegment: (LBRACKET expression? RBRACKET);

builtinTypeSpecifierParseable:
	BOOL
	| BVEC2
	| BVEC3
	| BVEC4
	| FLOAT16
	| F16VEC2
	| F16VEC3
	| F16VEC4
	| F16MAT2X2
	| F16MAT2X3
	| F16MAT2X4
	| F16MAT3X2
	| F16MAT3X3
	| F16MAT3X4
	| F16MAT4X2
	| F16MAT4X3
	| F16MAT4X4
	| FLOAT32
	| F32VEC2
	| F32VEC3
	| F32VEC4
	| F32MAT2X2
	| F32MAT2X3
	| F32MAT2X4
	| F32MAT3X2
	| F32MAT3X3
	| F32MAT3X4
	| F32MAT4X2
	| F32MAT4X3
	| F32MAT4X4
	| FLOAT64
	| F64VEC2
	| F64VEC3
	| F64VEC4
	| F64MAT2X2
	| F64MAT2X3
	| F64MAT2X4
	| F64MAT3X2
	| F64MAT3X3
	| F64MAT3X4
	| F64MAT4X2
	| F64MAT4X3
	| F64MAT4X4
	| INT8
	| I8VEC2
	| I8VEC3
	| I8VEC4
	| UINT8
	| U8VEC2
	| U8VEC3
	| U8VEC4
	| INT16
	| I16VEC2
	| I16VEC3
	| I16VEC4
	| UINT16
	| U16VEC2
	| U16VEC3
	| U16VEC4
	| INT32
	| I32VEC2
	| I32VEC3
	| I32VEC4
	| UINT32
	| U32VEC2
	| U32VEC3
	| U32VEC4
	| INT64
	| I64VEC2
	| I64VEC3
	| I64VEC4
	| UINT64
	| U64VEC2
	| U64VEC3
	| U64VEC4;

builtinTypeSpecifierFixed:
	VOID
	| ATOMIC_UINT
	| SAMPLER2D
	| SAMPLER3D
	| SAMPLERCUBE
	| SAMPLER2DSHADOW
	| SAMPLERCUBESHADOW
	| SAMPLER2DARRAY
	| SAMPLER2DARRAYSHADOW
	| SAMPLERCUBEARRAY
	| SAMPLERCUBEARRAYSHADOW
	| ISAMPLER2D
	| ISAMPLER3D
	| ISAMPLERCUBE
	| ISAMPLER2DARRAY
	| ISAMPLERCUBEARRAY
	| USAMPLER2D
	| USAMPLER3D
	| USAMPLERCUBE
	| USAMPLER2DARRAY
	| USAMPLERCUBEARRAY
	| SAMPLER1D
	| SAMPLER1DSHADOW
	| SAMPLER1DARRAY
	| SAMPLER1DARRAYSHADOW
	| ISAMPLER1D
	| ISAMPLER1DARRAY
	| USAMPLER1D
	| USAMPLER1DARRAY
	| SAMPLER2DRECT
	| SAMPLER2DRECTSHADOW
	| ISAMPLER2DRECT
	| USAMPLER2DRECT
	| SAMPLERBUFFER
	| ISAMPLERBUFFER
	| USAMPLERBUFFER
	| SAMPLER2DMS
	| ISAMPLER2DMS
	| USAMPLER2DMS
	| SAMPLER2DMSARRAY
	| ISAMPLER2DMSARRAY
	| USAMPLER2DMSARRAY
	| IMAGE2D
	| IIMAGE2D
	| UIMAGE2D
	| IMAGE3D
	| IIMAGE3D
	| UIMAGE3D
	| IMAGECUBE
	| IIMAGECUBE
	| UIMAGECUBE
	| IMAGEBUFFER
	| IIMAGEBUFFER
	| UIMAGEBUFFER
	| IMAGE1D
	| IIMAGE1D
	| UIMAGE1D
	| IMAGE1DARRAY
	| IIMAGE1DARRAY
	| UIMAGE1DARRAY
	| IMAGE2DRECT
	| IIMAGE2DRECT
	| UIMAGE2DRECT
	| IMAGE2DARRAY
	| IIMAGE2DARRAY
	| UIMAGE2DARRAY
	| IMAGECUBEARRAY
	| IIMAGECUBEARRAY
	| UIMAGECUBEARRAY
	| IMAGE2DMS
	| IIMAGE2DMS
	| UIMAGE2DMS
	| IMAGE2DMSARRAY
	| IIMAGE2DMSARRAY
	| UIMAGE2DMSARRAY;

structSpecifier: STRUCT IDENTIFIER? structBody;

structBody: LBRACE structMember+ RBRACE;

structMember:
	fullySpecifiedType structDeclarators += structDeclarator (
		COMMA structDeclarators += structDeclarator
	)* SEMICOLON;

structDeclarator: IDENTIFIER arraySpecifier?;

initializer:
	expression
	| LBRACE (
		initializers += initializer (COMMA initializers += initializer)* COMMA?
	)? RBRACE;

statement:
	compoundStatement
	| declarationStatement
	| expressionStatement
	| emptyStatement
	| selectionStatement
	| switchStatement
	| caseLabel
	| forStatement
	| whileStatement
	| doWhileStatement
	| jumpStatement
	| demoteStatement;

compoundStatement: LBRACE statement* RBRACE;

declarationStatement: declaration;

expressionStatement: expression SEMICOLON;

emptyStatement: SEMICOLON;

selectionStatement:
	attribute? IF LPAREN condition = expression RPAREN ifTrue = statement (
		ELSE ifFalse = statement
	)?;

iterationCondition:
	fullySpecifiedType name = IDENTIFIER ASSIGN_OP initializer;

switchStatement:
	attribute? SWITCH LPAREN condition = expression RPAREN compoundStatement;

caseLabel:
	CASE expression COLON	# valuedCaseLabel
	| DEFAULT COLON				# defaultCaseLabel;

whileStatement:
	attribute? WHILE LPAREN (
		condition = expression
		| initCondition = iterationCondition
	) RPAREN loopBody = statement;

doWhileStatement:
	attribute? DO loopBody = statement WHILE LPAREN condition = expression RPAREN SEMICOLON;

forStatement:
	attribute? FOR LPAREN (
		emptyStatement
		| expressionStatement
		| declarationStatement
	) (condition = expression | initCondition = iterationCondition)? SEMICOLON incrementer = expression? RPAREN
		loopBody = statement;

jumpStatement:
	CONTINUE SEMICOLON							# continueStatement
	| BREAK SEMICOLON								# breakStatement
	| RETURN expression? SEMICOLON	# returnStatement
	| DISCARD SEMICOLON							# discardStatement;

demoteStatement: DEMOTE SEMICOLON;
