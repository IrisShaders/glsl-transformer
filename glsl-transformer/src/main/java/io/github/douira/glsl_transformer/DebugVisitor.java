package io.github.douira.glsl_transformer;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import org.antlr.v4.runtime.tree.TerminalNode;

import io.github.douira.glsl_transformer.tree.ExtendedContext;
import io.github.douira.glsl_transformer.util.CompatUtil;

/**
 * The debug visitor prints the tree and some other information to the console.
 */
public class DebugVisitor extends GLSLParserBaseVisitor<Void> {
  private int maxDepth;

  /**
   * Creates a new debug visitor
   */
  public DebugVisitor() {
    this(Integer.MAX_VALUE);
  }

  /**
   * Creates a new debug visitor but with a maximum tree printing depth.
   * 
   * @param maxDepth The maximum depth to which the tree will be printed
   */
  public DebugVisitor(int maxDepth) {
    this.maxDepth = maxDepth;
  }

  @Override
  public Void visitChildren(RuleNode node) {
    var context = (ExtendedContext) node.getRuleContext();

    var depth = (context.getPayload()).depth();

    if (depth > maxDepth) {
      return null;
    }

    // indent with the tree depth
    var prefix = CompatUtil.repeat(". ", depth);
    int n = node.getChildCount();

    // print the name of the node
    var name = node.getPayload().getClass().getSimpleName();
    if (n > 1 || name.indexOf("Expression") == -1) {
      System.out.println(prefix + "---" + name);
    }

    for (int i = 0; i < n; i++) {
      ParseTree c = node.getChild(i);

      if (n > 1) {
        System.out.println(prefix + i + "/" + n + " in " + name);
        System.out.println(prefix + c.getText());
      }

      if (c instanceof TerminalNode terminal) {
        System.out.println(prefix + terminal.getSymbol().getText());
      }

      c.accept(this);
    }

    return null;
  }
}
