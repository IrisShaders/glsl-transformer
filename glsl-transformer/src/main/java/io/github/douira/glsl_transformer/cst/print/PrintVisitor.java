package io.github.douira.glsl_transformer.cst.print;

import java.util.LinkedList;

import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.misc.Interval;
import org.antlr.v4.runtime.tree.*;

import com.github.bsideup.jabel.Desugar;

import io.github.douira.glsl_transformer.cst.node.UnparsableCSTNode;
import io.github.douira.glsl_transformer.cst.token_filter.TokenFilter;
import io.github.douira.glsl_transformer.tree.ExtendedContext;

/**
 * The print visitor visits the parse tree and reprints it while preserving the
 * position and content of hidden tokens. This means it preserves all newlines
 * even if the parsed nodes that contains them was removed. This is useful
 * because it doesn't necessarily change the line numbers during patching.
 * 
 * Local roots in the parse tree tell print visitor what changes have been made
 * to the original parse tree and from which alternative token streams it has to
 * read when it prints newly inserted nodes (=local roots) and their subtrees.
 */
public class PrintVisitor extends AbstractParseTreeVisitor<Void> {
  /**
   * The attributed interval is used to attribute intervals in the print list to
   * local root nodes. Which root node an interval belongs to is used during
   * printing to read from the token stream for which the interval was
   * constructed.
   */
  @Desugar
  private static record AttributedInterval(ExtendedContext localRoot, Interval interval) {
  }

  private final LinkedList<Object> printItems = new LinkedList<>();
  private Interval cachedInterval;
  private ExtendedContext currentRoot;

  private PrintVisitor() {
  }

  private static boolean inInterval(Interval interval, int el) {
    return interval.a <= el && interval.b >= el;
  }

  /**
   * Prints the given parse tree that references the given token stream. Sets up
   * the given node as a root node if that hasn't happened already.
   * 
   * If the given token filter is not {@code null}, it will be used to check if
   * each otherwise qualified token should be printed.
   * 
   * @param rootTokenStream The token stream for the parse tree
   * @param tree            The parse tree to print
   * @param tokenFilter     An additional token filter to check before printing
   *                        each otherwise qualified token
   * @return The printed parse tree that includes the results of the
   *         transformations
   */
  public static String printTree(
      BufferedTokenStream rootTokenStream, ExtendedContext tree, TokenFilter<?> tokenFilter) {
    tree.makeLocalRoot(rootTokenStream);
    return new PrintVisitor()
        .visitAndJoin(rootTokenStream, tree, tree.getFullSourceInterval(), tokenFilter);
  }

  /**
   * Prints the given parse tree that references the given token stream. Sets up
   * the given node as a root node if that hasn't happened already.
   * 
   * @see #visitAndJoin(BufferedTokenStream, ExtendedContext, Interval,
   *      TokenFilter)
   * @see #printTree(BufferedTokenStream, ExtendedContext, TokenFilter)
   * 
   * @param rootTokenStream The token stream for the parse tree
   * @param tree            The parse tree to print
   * @return The printed parse tree that includes the results of the
   *         transformations
   */
  public static String printTree(
      BufferedTokenStream rootTokenStream, ExtendedContext tree) {
    return printTree(rootTokenStream, tree, null);
  }

  /**
   * This is the lowest level printing method that should be used publicly. It
   * takes a parse tree and a token stream and prints it. Additionally an interval
   * is given that determines which tokens of the string should be printed.
   * 
   * @see #printTree(BufferedTokenStream, ExtendedContext)
   * @param rootTokenStream The token stream for the parse tree
   * @param rootNode        The parse tree to print
   * @param bounds          An token index interval of the tokens that should be
   *                        included in the printed string
   * @param tokenFilter     An additional token filter for checking if tokens
   *                        should be printed
   * @return The printed parse tree that is within the bounds and contains the
   *         results of the transformations applied to the tree
   */
  private String visitAndJoin(BufferedTokenStream rootTokenStream,
      ExtendedContext rootNode, Interval bounds, TokenFilter<?> tokenFilter) {
    // visit the whole tree and accumulate tokens and intervals
    currentRoot = rootNode;
    visit(rootNode);

    if (tokenFilter != null) {
      tokenFilter.resetState();
    }

    // convert the list of tokens and intervals into just tokens,
    // and then into their strings
    var builder = new StringBuilder(512); // guessing
    for (var printItem : printItems) {
      if (printItem instanceof String literal) {
        builder.append(literal);
      } else if (printItem instanceof AttributedInterval attributedInterval) {
        var interval = attributedInterval.interval();
        var localRoot = attributedInterval.localRoot();
        var omissionSet = localRoot.getLocalRootTokenOmissions();

        for (var token : localRoot.getTokenStream().getTokens(interval.a, interval.b)) {
          // don't print EOF, only print the tokens in side the printing bounds,
          // but always allow inserted nodes, only print non-omitted tokens
          var tokenIndex = token.getTokenIndex();
          if (token.getType() != Lexer.EOF
              && (tokenIndex == -1
                  || (localRoot != rootNode || inInterval(bounds, tokenIndex))
                      && omissionSet.tokenNotOmitted(token)
                      && (tokenFilter == null || tokenFilter.isTokenAllowed(token)))) {
            builder.append(token.getText());
          }
        }
      } else {
        throw new AssertionError(
            "A wrong type of object was inserted into the printItems! Only String and AttributedInterval are allowed.");
      }
    }
    return builder.toString();
  }

  /**
   * Adds an interval to the current list of intervals to print. The interval is
   * attributed to the current local root node.
   * 
   * @param a The left bound of the interval
   * @param b The right bound of the interval
   */
  private void addInterval(int a, int b) {
    // doing this check here saves an object construction if the interval is invalid
    // (and it often is invalid)
    if (a > b || a < 0 || b < 0) {
      return;
    }

    // cache a single interval since often the same interval is added repeatedly
    Interval interval;
    if (cachedInterval != null && a == cachedInterval.a && b == cachedInterval.b) {
      interval = cachedInterval;
    } else {
      interval = Interval.of(a, b);
      cachedInterval = interval;
    }

    addInterval(interval);
  }

  /**
   * Adds an interval to the current list of intervals to print. The interval is
   * attributed to the current local root node. If possible, this will join the
   * interval with the last added interval in order to reduce memory usage and
   * overhead during printing. Oftentimes the intervals are adjacent and share the
   * same local root, which makes them joinable. If an interval is empty, nothing
   * needs to be done with it since it can't contribute any tokens for printing.
   * 
   * @param newInterval The interval to add to the print list
   */
  private void addInterval(Interval newInterval) {
    if (newInterval.length() == 0) {
      return;
    }

    // join the given interval onto the last interval if possible without holes
    if (!printItems.isEmpty()) {
      var lastPrintItem = printItems.getLast();

      // only check for combining if last item is also an interval
      if (lastPrintItem instanceof AttributedInterval lastPrintInterval) {
        var lastInterval = lastPrintInterval.interval();
        if (lastInterval != null
            && currentRoot == lastPrintInterval.localRoot()
            && (!lastInterval.disjoint(newInterval)
                || lastInterval.adjacent(newInterval))) {
          if (lastInterval.properlyContains(newInterval)) {
            return;
          }

          printItems.removeLast();
          newInterval = lastInterval.union(newInterval);
        }
      }
    }

    printItems.add(new AttributedInterval(currentRoot, newInterval));
  }

  /**
   * Adds a string as a literal to be printed to the print list
   * 
   * @param literal
   */
  private void addLiteral(String literal) {
    if (literal == null || literal.isEmpty()) {
      return;
    }

    printItems.add(literal);
  }

  /**
   * {@inheritDoc}
   * 
   * The visitor method that the print visitor overrides in order to collect each
   * node's intervals. The current local root is updated when a local root node is
   * found. Any created intervals for printing are attributed to the current local
   * root in order to know which token stream the interval's indexes refer to.
   * 
   * @implNote intervals that cover tokens that are not part of children (which
   *           are only hidden tokens like whitespace because terminal nodes are
   *           children too) are moved after inserted nodes, which are local
   *           roots, through this process. this happens because fetchNext is not
   *           updated for local roots which causes the next non-local-root child
   *           (or at the after end of the child list) to add the whole interval
   *           covering the child after the child itself has been visited.
   */
  @Override
  public Void visitChildren(RuleNode node) {
    final var context = (ExtendedContext) node.getRuleContext();
    final var superInterval = context.getLargestSourceInterval();
    var lastWasUnparsableASTNode = false;

    // keeps track of the token index that needs to be processed next
    var fetchNext = superInterval.a;

    // iterate all children of this node. This uses a for loop so that previous and
    // next children can be accessed
    if (context.children != null) {
      var childrenLength = context.children.size();
      for (int i = 0; i < childrenLength; i++) {
        var child = context.children.get(i);

        // prettify unparsable AST nodes
        if (child instanceof UnparsableCSTNode) {
          // insert a newline before each group of unparsable ast nodes.
          // line preservation doesn't matter here since it's being broken anyways
          if (!lastWasUnparsableASTNode && ((UnparsableCSTNode) child).doNewlineInsertion()) {
            addLiteral("\n");
            lastWasUnparsableASTNode = true;
          }

          // unparsable ast nodes have no source interval (length 0)
          // they will follow the path into opening detection
        } else {
          lastWasUnparsableASTNode = false;
        }

        // local root node handling requires updating the current root node from which
        // the openings are retrieved and that is annotated on the intervals
        if (child instanceof ExtendedContext childNode && childNode.isLocalRoot()) {
          // set as new current root
          var previousRoot = currentRoot;
          currentRoot = childNode;

          child.accept(this);
          currentRoot = previousRoot;
        } else {
          // handle everything else (regular non-terminal and terminal nodes)
          // interval before the current child
          var childInterval = child.getSourceInterval();
          if (childInterval.length() > 0) {
            addInterval(fetchNext, childInterval.a - 1);
            child.accept(this);

            // prevent an empty child interval from messing up fetchNext by being negative
            // (fetchNext is only updated if childInterval is not empty)
            fetchNext = childInterval.b + 1;
          } else {
            // detect openings left by removed nodes for new nodes that have no source
            // interval (e.g. unparsable ast nodes but also other such nodes).
            // The interval has to be invalid, not just empty since some normal nodes have
            // empty intervals when they represent empty lists (like empty function
            // parameter lists)
            if (childInterval == Interval.INVALID) {
              var previousChildEnd = i > 0
                  ? context.children.get(i - 1).getSourceInterval().b
                  : -1;

              // find the next opening that this node can be inserted into
              // but make sure it's between the end of the last child, the beginning of the
              // next child and not before the current fetchNext token position. (if fetchNext
              // is corrupted, the printer will wrongly output tokens multiple times)
              var nextOpening = currentRoot.getLocalRootOpenings().higher(previousChildEnd);
              if (nextOpening != null && nextOpening >= fetchNext) {
                var nextChildStart = i < childrenLength - 1
                    ? context.children.get(i + 1).getSourceInterval().a
                    : -1;
                if (nextChildStart == -1 || nextOpening < nextChildStart) {
                  // add the interval of tokens up to the next opening and advance the nextIndex
                  // so that the whitespace before the opening is placed before the inserted node
                  addInterval(fetchNext, nextOpening);
                  fetchNext = nextOpening + 1;
                }
              }
            }
            child.accept(this);
          }
        }
      }
    }

    // interval after the last child
    addInterval(fetchNext, superInterval.b);
    return null;
  }

  @Override
  public Void visitTerminal(TerminalNode node) {
    // empty terminal nodes have an empty source interval and will have no effect
    addInterval(node.getSourceInterval());

    // if this is an unparsable AST node, add the literal it produces
    if (node instanceof UnparsableCSTNode astNode) {
      addLiteral(astNode.getText());
    }
    return null;
  }
}
