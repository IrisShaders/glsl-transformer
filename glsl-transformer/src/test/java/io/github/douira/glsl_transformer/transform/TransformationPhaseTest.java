package io.github.douira.glsl_transformer.transform;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ArgumentsSource;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import io.github.douira.glsl_transformer.GLSLParser;
import io.github.douira.glsl_transformer.IntegratedTest;
import io.github.douira.glsl_transformer.TestCaseProvider;
import io.github.douira.glsl_transformer.GLSLParser.TranslationUnitContext;
import io.github.douira.glsl_transformer.generic.PrintVisitor;

@ExtendWith({ SnapshotExtension.class })
public class TransformationPhaseTest extends IntegratedTest {
  private Expect expect;

  class EmptyPhase extends TransformationPhase {
  }

  @BeforeAll
  static void setupInput() {
    readInput(TransformationPhaseTest.class);
  }

  @Test
  void testCompilePath() {
    wrapRunTransform(new EmptyPhase() {
      @Override
      protected void init() {
        var path = compilePath("/translationUnit/externalDeclaration");
        assertEquals(path.evaluate(tree).size(), tree.getChildCount() - 2, "It should compile a functioning xpath");
      }
    });
  }

  @Test
  void testCompilePattern() {
    wrapRunTransform(new EmptyPhase() {
      @Override
      protected void init() {
        var pattern = compilePattern("varying <type:typeSpecifier> varyVec;",
            GLSLParser.RULE_externalDeclaration);
        var match = pattern.match(tree.getChild(5));
        assertTrue(match.succeeded(), "It should compile a functioning pattern");
        assertEquals("vec2", match.get("type").getText(), "It should compile a functioning pattern");
      }
    });

  }

  @Test
  void testCreateLocalRoot() {

  }

  @Test
  void testFindAndMatch() {

  }

  @Test
  void testGetSiblings() {
    assertEquals(tree.children, TransformationPhase.getSiblings(tree.versionStatement()),
        "It should find the siblings of a node");
  }

  @Test
  void testInjectExternalDeclaration() {

  }

  @ParameterizedTest
  @ArgumentsSource(TestCaseProvider.class)
  @SnapshotName("testInjectNode")
  void testInjectNode(String scenario, String input) {
    //TODO: test case loader with snapshot name,
    //TODO: fix missing semicolon in empty test
    setupParsingWith(input);
    wrapRunTransform(new RunPhase() {
      @Override
      protected void run(TranslationUnitContext ctx) {
        injectExternalDeclaration("foo;\n", InjectionPoint.BEFORE_VERSION);
      }
    });

    var output = PrintVisitor.printTree(tokenStream, tree);
    expect.scenario(scenario).toMatchSnapshot(input, "=====", output);
  }

  @Test
  void testIsActive() {
    EmptyPhase phase = new EmptyPhase();
    assertTrue(phase.isActive(), "It should always be active");
  }

  @Test
  void testRemoveNode() {

  }

  @Test
  void testReplaceNode() {

  }

  @Test
  void testSetParent() {
    wrapRunTransform(new RunPhase() {
      @Override
      protected void run(TranslationUnitContext ctx) {
        assertEquals(parser, getParser(),
            "It should return the previously set parser inside the phase collector");
      }
    });
  }
}
