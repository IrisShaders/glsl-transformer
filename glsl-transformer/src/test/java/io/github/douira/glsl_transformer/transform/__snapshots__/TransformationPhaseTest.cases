§testInjectNode empty:

§testInjectNode unparsed_prefix:
//comment prefix
#version 5432

§testInjectNode single_version:
#version 5432

§testInjectNode single_extension:
#extension extension_name: require

§testInjectNode single_declaration:
varying vec2 varyVec;

§testInjectNode single_layout:
layout(location = 0) out vec3 layoutOne;

§testInjectNode single_pragma:
#pragma foo

§testInjectNode single_function:
void main() {}

§testInjectNode full:
#version 5432

#extension extension_name: require

#pragma foo

varying vec2 varyVec;

void functionDef(const in float param) {
  functionCall();
}

§testInjectNode full_reverse:
#version 5432

void functionDef(const in float param) {
  functionCall();
}

varying vec2 varyVec;

#pragma foo

#extension extension_name: require

§testInjectNode version_with_pragma:
#version 5432

#pragma foo

§testInjectNode version_with_extension:
#version 5432

#extension extension_name: require
