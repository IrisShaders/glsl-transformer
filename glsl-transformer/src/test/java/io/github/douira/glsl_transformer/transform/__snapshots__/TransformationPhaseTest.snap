testInjectNode[2]=[
bar;
=====
foo;bar;
]


testInjectNode[empty]=[

=====
foo;
]


testInjectNode[simple]=[
#version 5432
bar;
=====
foo;#version 5432
bar;
]